### DISCUSSION TITLE
Berry read binary file

### DISCUSSION THREAD

**Discussion by DetlefR** [arendst/Tasmota#16382] [Q&A] [answered]
Hello,

I try to read a binary file with a pattern for a WS2812 LED strip
I have tried pixel=ledfile.read(3) to read the 3 bytes for a LED. But unfortunately the result is not an integer or byte array, it is a string.
I found no way to convert pixel easy to integer.
At the moment, I use this terrible code.
pixel=string.byte(ledfile.read(1))
pixel+=string.byte(ledfile.read(1))*256
pixel+=string.byte(ledfile.read(1))*65536

Does anyone know a better way

**Comment by s-hadinger** [comment, marked as answer]
```
pixel.readbytes(3)
```

I realize that the `file` class is missing from the documentation. I will have to fix that.

**Reply by sfromis** [reply]
While the outcome from readbytes is not an integer either, the bytes class does have methods for getting numbers (and other things) from the raw data. However, 3-byte numbers are not directly supported there, either....
https://github.com/berry-lang/berry/wiki/Chapter-7#bytes-class

If you already have a string, you can also construct a bytes object:
`var rawbytes = bytes().fromstring(str)`

**Reply by barbudor** [reply]
I imagine the file is created especially to be parsed by Berry by another desktop software
So the best is probably to write the file in a format that is easy to read from Berry
You could just write data as a text file with the text representation of the integer made of the 3 bytes on each line

**Reply by DetlefR** [reply]
Hello,

Thanks 
`var rawbytes = bytes().fromstring(str)`
is that for what I was looking for.

@barbudor your proposal is also possible. But I prefere pur binary. Is smaler and faster.

**Reply by s-hadinger** [reply]
Be careful with strings. They don't always handle well Null chars if any

**Reply by sfromis** [reply]
When it is from a file, better use the method suggested by s-hadinger, readbytes directly from the file, without going through a string (which is just a bad detour for binary data).

More generally, for neopixel data formats, I'd prefer 4 bytes per pixel, as some neopixel strips have white in addition to RGB. And the bytes object has nice methods (get/geti) for taking 4-byte numbers out of the raw byte buffer.

**Reply by DetlefR** [reply]
Thanks for your ideas.
I have made a mix from it
`var starttime=tasmota.millis()

var ledfile=open("/test1.led")

var pixels=bytes(ledfile.read())

ledfile.close()

print(tasmota.millis()-starttime)

var pos=0

var lbuf=stripe.pixels_buffer()

while pos<lbuf.size()

 lbuf[pos]=pixels[pos]

 pos+=1

end  

stripe.dirty()

stripe.show()

print(tasmota.millis()-starttime) `

I use a file with hex values. But not line by line, but one after the other.
`4bfd21e7a2d0f85898c463081ffa80466fd...`
What I now looking for is a replacement for the while loop. Open file, read and close needs 25 millis. The loop another 200.

**Reply by sfromis** [reply]
Faster:
```
for pos:0..lbuf.size()-1
  lbuf[pos]=pixels[pos]
end
```
Still unsafe in case of length mismatch from file to pixels buffer but that's another question....

(Text blocks is triple backticks before and after)

**Reply by DetlefR** [reply]
Yes is faster. From 230 to 150 millis.
You are right. The lenght of the file should fit the amount of Led. Or the number of characters to read is fixed.
```
var pixels=bytes(ledfile.read(pixel_count * pixel_size * 2))
```
Thanks for the tip with the code block

**Reply by sfromis** [reply]
Much faster (simplified for multiplum of 4):
```
for pos:0..(lbuf.size()-1)/4
  lbuf.set(pos<<2,pixels.get(pos<<2,4),4)
end
```
Did not really check if the ending condition is perfect, but you get the point....

**Reply by sfromis** [reply]
If it works, way faster:
```
var ledfile=open("/test1.led")
var pixels=ledfile.read()
ledfile.close()
var lbuf=stripe.pixels_buffer()
lbuf.fromhex(pixels)
```
I've not checked if this method actually works with the neopixel buffer byte array.

**Reply by s-hadinger** [reply]
It should work as long as the length match exactly

**Reply by DetlefR** [reply]
@sfromis 
```
var starttime=tasmota.millis()
var ledfile=open("/test1.led")
var pixel_count=stripe.pixel_count()
var pixel_size=stripe.pixel_size()
var pixels=ledfile.read(pixel_count * pixel_size * 2)
ledfile.close()
var lbuf=stripe.pixels_buffer()
lbuf.fromhex(pixels)
stripe.dirty()
stripe.show()
print(tasmota.millis()-starttime) 
```
Result is beteen 42 and 50 millis. 
I think faster is not possible.

**Reply by sfromis** [reply]
You could trim it a minuscule bit by using values directly, instead of putting it in a variable to be used only once, but I'm not sure that you'd get a clear result, especially when time is already short. Another tiny tweak would be `< 1` (shift) instead of the multiplication `* 2`. You already saw the benefit of cutting out extra variable-handling steps in the first improvement, but of course that only became significant due to it being in a loop.
OTOH, I'd probably like to add a sanity check to bail out in case the number of pixels from the file does not match the expected.

**Comment by DetlefR** [comment]
I have begun to write what I actually wanted to. A class to rotate a pattern over a LED panel. She is working. Almost.
Instead of move the pattern row by row, it makes only ever second step. The test file contains only a few LED with color. The most are black/dark. For debugging, I have inserted the print statement. As you can see, the first hexbyte (pixles[0]) is changed on every step. But the LED buffer "ignore" every second step. 
I'm a little confused.

```
print(self.lbuf[0],pixel,self.col_offset,currentpos,rest,tasmota.millis()-starttime,tasmota.time_str(tasmota.millis()/1000))
196 c4 10 480 0 50 1970-01-01T00:02:08
196 00 11 528 0 51 1970-01-01T00:02:09
0 00 12 576 0 50 1970-01-01T00:02:10
...
0 00 24 1152 288 58 1970-01-01T00:02:22
0 34 25 1200 336 52 1970-01-01T00:02:23
```

```
import string
class ROTATE 
 var stripe
 var rows
 var filename
 var pixels
 var isvalid
 var pixel_count
 var pixel_size
 var col_offset
 var lbuf
 var running
 var displaycol
 var intcount
 
 def init(stripe,filename,rows)
  self.stripe=stripe   #Leds class
  self.filename=filename #Name and path of pattern file
  self.rows=rows       #Number of rows 
  self.pixel_count=stripe.pixel_count()
  self.pixel_size=stripe.pixel_size()
  self.lbuf=stripe.pixels_buffer()
  var ledfile=open(self.filename)
  self.pixels=ledfile.read()
  ledfile.close()
  self.isvalid=(size(self.pixels) % (rows*self.pixel_size*2))==0 # file size must a multiple of (rows*self.pixel_size*2)
  self.displaycol=size(self.pixels)/(rows*self.pixel_size*2)
  self.col_offset=0
  self.running=false
  self.intcount=0
  self.start()
 end
 
 def stop()
  if self.running
   tasmota.remove_driver(self)
  end
  self.running=false  
 end
 
 def start()
  if !self.running && self.isvalid
   tasmota.add_driver(self)
   self.running=true   
  end 
 end
 
 def every_second()
  if !self.isvalid return end
  var starttime=tasmota.millis() 
  var currentpos=self.col_offset*self.rows*self.pixel_size*2
  var ledfile=open(self.filename)
  ledfile.seek(currentpos)
  var pixels=ledfile.read(self.lbuf.size()*2)
  var rest=(self.lbuf.size()*2)-size(pixels)
  if rest>0
   ledfile.seek(0)
   pixels+=ledfile.read(rest)
  end
  ledfile.close()
  self.lbuf.fromhex(pixels)
  self.stripe.dirty()
  self.stripe.show()
  var pixel=string.split(pixels,2)[0]
  print(self.lbuf[0],pixel,self.col_offset,currentpos,rest,tasmota.millis()-starttime,tasmota.time_str(tasmota.millis()/1000))
  self.col_offset=self.col_offset>=(self.displaycol-1) ? 0 : self.col_offset+1
 end
end

var stripe=Leds(256,17)
var rotate=ROTATE(stripe,'/test2.led',8)

```
[test2.txt](https://github.com/arendst/Tasmota/files/9472221/test2.txt)

**Reply by sfromis** [reply]
> But the LED buffer "ignore" every second step.

What do you mean by that? To me it is not clear where specifically you see your expectations broken....

**Reply by s-hadinger** [reply]
I'm curious what's the overhead of opening and closing the file at each iteration. You can keep it open. But it's maybe negligible

**Reply by DetlefR** [reply]
@sfromis 
I read every second the pattern file new. With a offset of one column "rows*self.pixel_size*2" and copy it LED buffer "self.lbuf.fromhex(pixels)". For debugging, I have made insert the print statement. The firts value is the first byte in LED buffer. The second the two first byte from readbuffer "pixels". The third is the column offset and the fourth the offset in pattern file.
The first line in print output shows that the first led buffer contains "196" that is the correct expression of "c4" from pattern file.
```196 c4 10 480 0 50 1970-01-01T00:02:08```
The next line, one second later, shows the next step. The file offset is incremented of 48 bytes "480->528". The "redbuffer" is "00" here but the "ledbuffer" ist still "196". 
```196 00 11 528 0 51 1970-01-01T00:02:09```
The last line from print output shows "34" in "readbuffer" but the "ledbuffer" is "0" from step before.

And this is so every time.
That's what I not understand.

**Reply by sfromis** [reply]
I've not gone into implementation internals, but with the minuscule amount of debugging you did, I speculate that your expectation of it being safe to store `stripe.pixels_buffer()` in `self.lbuf` in the initialization only may be risky. Things could plausibly be reallocated for whatever reasons.
